src/masterless.py

```diff
@@ -80,11 +60,62 @@ def set_versions(stackname, *repolist):
 
         arglist.append((repo, known_formula_map[repo], rev))
 
-    if not arglist:
+    # test given revisions actually exist in formulas
+    for name, _, revision in arglist:
+        path = join(config.PROJECT_PATH, "cloned-projects", name)
+        if not os.path.exists(path):
+            LOG.warn("couldn't find formula %r locally, revision check skipped", path)
+            continue
+
+        with lcd(path), settings(warn_only=True):
+            ensure(local("git fetch --quiet").succeeded, "failed to fetch remote refs for %s" % path)
+            ensure(local("git cat-file -e %s^{commit}" % revision).succeeded, "failed to find ref %r in %s" % (revision, name))
+
+    return arglist
+
+#
+#
+
+@task
+@requires_project
+@requires_master_server_access
+def launch(pname, instance_id=None, alt_config='standalone', *repolist):
+    pdata = project.project_data(pname)
+    # ensure given alt config has masterless=True
+    ensure(pdata['aws-alt'], "project has no alternate configurations")
+    ensure(alt_config in pdata['aws-alt'], "unknown alt-config %r" % alt_config)
+    ensure(pdata['aws-alt'][alt_config]['ec2']['masterless'], "alternative configuration %r has masterless=False" % alt_config)
+    repolist = parse_validate_repolist(pdata, *repolist)
+
+    import cfn
+    cfn.launch(pname, instance_id, alt_config, formula_revisions=repolist)
+
+
+@task
+@requires_aws_stack
+@requires_master_server_access
+@requires_masterless
+def update(stackname):
+    import cfn
+    return cfn.update(stackname)
```

Comment:
do we need a separate `masterless.update` task, given it's just delegating? Or the annotations are doing very smart things?

Reply:
no, I don't think we need it. It was an early effort to provide a similar interface to cfn.py, but now it's not doing much besides being noisy.